src/Build/BackEnd/Components/Communications/NodeEndpointInProc.cs

```diff
@@ -332,11 +333,8 @@ private void EnqueuePacket(INodePacket packet)
             ErrorUtilities.VerifyThrow(null != _packetQueue, "packetQueue is null");
             ErrorUtilities.VerifyThrow(null != _packetAvailable, "packetAvailable is null");
 
-            lock (_packetQueue)
```

Comment:
I wonder if weird check-then-act style issues can appear here. Before, enqueing and setting the event happened atomically, but now they don't anymore.

One possible new interleaving is when the producer enqueues a packet but before it gets to set the event, the consumer drains the queue from a previous event signal. Then the producer signals the event, and the consumer does a dry run on the empty queue. This appears to be harmless.

I wonder though if there's an interleaving where the consumer misses notifications and packets get stuck on the queue. As far as I can tell it shouldn't happen, but another confirmation would be nice. Would have been nice if ConcurrentQueue offered an ElementAdded event.

Reply:
With lock here the same situation is still possible. If we have 2+ threads put packets, consumer also might drain the whole queue.